### PR TITLE
1st_version of the github_auto_add

### LINKED ISSUE #185: Auto-add github resources
Similar to the [automatic intake of meta-data from zenodo](https://github.com/NFDI4BIOIMAGE/training/blob/main/scripts/auto-add-zenodo-entry.py), we may want to have this for github-resources. The [python github-API](https://github.com/PyGithub/PyGithub) might be useful for this.

If a new issue is created, and the body is only a github link, we could read the meta data of the repository and send a pull-request to add it to our list of resource.

### PR DESCRIPTION
**Pull request by SeverusYixin**
Hi @haesleinhuepf, would you mind helping me review it?

closes #185
closes #246

### CODE REVIEW

File: scripts/auto-add-github-resources.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    """
+    Main function to retrieve metadata from a GitHub repository and append it to a YAML file. 
+    Creates a pull request with the updated data.
+
+    Arguments:
+    - repository (str): GitHub repository name.
+    - issue (int): Issue number containing the GitHub link.
+
+    Example:
+    python auto_add_github_entry.py "username/repo" 123
```
Comment (haesleinhuepf) [inline scripts/auto-add-github-resources.py:17]:
```suggestion
    python auto_add_github_entry.py username/repo 123
```
I presume this is correct?

**Comment by SeverusYixin** [comment]
Why can I not write this link like you did "https://github.com/NFDI4BIOIMAGE/training/blob/2f17431f8fbd1ee4ffb205eb6fe26474a04ec41a/scripts/auto-add-zenodo-entry.py#L1"?

**Review by haesleinhuepf — CHANGES_REQUESTED** [review]
Hi @SeverusYixin ,

awesome, thanks for this! Two things: 
1. Before sending a PR, please check if the tool works as you describe it in the documentation.
2. I just tested it and received an error message. That's the command I ran:
```
python scripts/auto-add-github-resources.py nfdi4bioimage/training 236
```

And that's the error:
```
-> get_conversation_on_issue(nfdi4bioimage/training, 236)
Request GET /repos/rmassei/ymia_python_event_series_material/collaborators?per_page=1 failed with 403: Forbidden
Traceback (most recent call last):
  File "C:\structure\code\training\scripts\auto-add-github-resources.py", line 139, in <module>
    main()
  File "C:\structure\code\training\scripts\auto-add-github-resources.py", line 34, in main
    github_data_dict = complete_github_data(github_repo_url)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\structure\code\training\scripts\auto-add-github-resources.py", line 97, in complete_github_data
    if collaborators.totalCount > 0:
       ^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\github\PaginatedList.py", line 192, in totalCount
    headers, data = self.__requester.requestJsonAndCheck(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\github\Requester.py", line 550, in requestJsonAndCheck
    return self.__check(*self.requestJson(verb, url, parameters, headers, input, self.__customConnection(url)))
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\github\Requester.py", line 715, in requestJson
    return self.__requestEncode(cnx, verb, url, parameters, headers, input, encode)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\github\Requester.py", line 812, in __requestEncode
    status, responseHeaders, output = self.__requestRaw(cnx, verb, url, requestHeaders, encoded_input)
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\github\Requester.py", line 846, in __requestRaw
    response = cnx.getresponse()
               ^^^^^^^^^^^^^^^^^
  File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\github\Requester.py", line 198, in getresponse
    r = verb(
        ^^^^^
  File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\requests\sessions.py", line 602, in get
    return self.request("GET", url, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\requests\sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\requests\sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\requests\adapters.py", line 667, in send
    resp = conn.urlopen(
           ^^^^^^^^^^^^^
  File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\urllib3\connectionpool.py", line 934, in urlopen
    retries = retries.increment(method, url, response=response, _pool=self)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\github\GithubRetry.py", line 187, in increment
    raise Requester.createException(response.status, response.headers, content)  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
github.GithubException.GithubException: 403 {"message": "Must have push access to view repository collaborators.", "documentation_url": "https://docs.github.com/rest/collaborators/collaborators#list-repository-collaborators", "status": "403"}
```

Let me know if I should try again!

And thanks again for working on this. This will be cool to have!

Best,
Robert

File: scripts/auto-add-github-resources.py (hunk 1 above)
Comment (SeverusYixin) [reply scripts/auto-add-github-resources.py]: I have deleted it. I wanted to test it locally or on my github, but it seems I'm not allowed to.

**Comment by SeverusYixin** [comment]
> Hi @SeverusYixin ,
> 
> awesome, thanks for this! Two things:
> 
> 1. Before sending a PR, please check if the tool works as you describe it in the documentation.
> 2. I just tested it and received an error message. That's the command I ran:
> 
> ```
> python scripts/auto-add-github-resources.py nfdi4bioimage/training 236
> ```
> 
> And that's the error:
> 
> ```
> -> get_conversation_on_issue(nfdi4bioimage/training, 236)
> Request GET /repos/rmassei/ymia_python_event_series_material/collaborators?per_page=1 failed with 403: Forbidden
> Traceback (most recent call last):
>   File "C:\structure\code\training\scripts\auto-add-github-resources.py", line 139, in <module>
>     main()
>   File "C:\structure\code\training\scripts\auto-add-github-resources.py", line 34, in main
>     github_data_dict = complete_github_data(github_repo_url)
>                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\structure\code\training\scripts\auto-add-github-resources.py", line 97, in complete_github_data
>     if collaborators.totalCount > 0:
>        ^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\github\PaginatedList.py", line 192, in totalCount
>     headers, data = self.__requester.requestJsonAndCheck(
>                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\github\Requester.py", line 550, in requestJsonAndCheck
>     return self.__check(*self.requestJson(verb, url, parameters, headers, input, self.__customConnection(url)))
>                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\github\Requester.py", line 715, in requestJson
>     return self.__requestEncode(cnx, verb, url, parameters, headers, input, encode)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\github\Requester.py", line 812, in __requestEncode
>     status, responseHeaders, output = self.__requestRaw(cnx, verb, url, requestHeaders, encoded_input)
>                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\github\Requester.py", line 846, in __requestRaw
>     response = cnx.getresponse()
>                ^^^^^^^^^^^^^^^^^
>   File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\github\Requester.py", line 198, in getresponse
>     r = verb(
>         ^^^^^
>   File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\requests\sessions.py", line 602, in get
>     return self.request("GET", url, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\requests\sessions.py", line 589, in request
>     resp = self.send(prep, **send_kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\requests\sessions.py", line 703, in send
>     r = adapter.send(request, **kwargs)
>         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\requests\adapters.py", line 667, in send
>     resp = conn.urlopen(
>            ^^^^^^^^^^^^^
>   File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\urllib3\connectionpool.py", line 934, in urlopen
>     retries = retries.increment(method, url, response=response, _pool=self)
>               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\rober\miniforge3\envs\devbio-napari-env\Lib\site-packages\github\GithubRetry.py", line 187, in increment
>     raise Requester.createException(response.status, response.headers, content)  # type: ignore
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> github.GithubException.GithubException: 403 {"message": "Must have push access to view repository collaborators.", "documentation_url": "https://docs.github.com/rest/collaborators/collaborators#list-repository-collaborators", "status": "403"}
> ```
> 
> Let me know if I should try again!
> 
> And thanks again for working on this. This will be cool to have!
> 
> Best, Robert

The problem is that I can not test it, it told me the scads.ai clould refuse me XD

**Comment by haesleinhuepf** [comment]
What is "scads.ai clould" in this context? What was the exact error message?

**Comment by haesleinhuepf** [comment]
The error above is not about our repository:
> Request GET /repos/rmassei/ymia_python_event_series_material/collaborators?per_page=1 failed with 403: Forbidden
> ...
> Must have push access to view repository collaborators.

The error happens when accessing the other repository. I'd say we will not be able to read out collaborators from remote repositories.

**Comment by SeverusYixin** [comment]
> The error above is not about our repository:
> 
> > Request GET /repos/rmassei/ymia_python_event_series_material/collaborators?per_page=1 failed with 403: Forbidden
> > ...
> > Must have push access to view repository collaborators.
> 
> The error happens when accessing the other repository. I'd say we will not be able to read out collaborators from remote repositories.

Yes, I used the code snippet to check locally that the link to our repository is no problem, it show me as PS C:\Users\huang\Documents\GitHub\training\scripts> & 'c:\Users\huang\AppData\Local\Programs\ Python\Python312\python. exe' 'c:\Users\huang\.vscode\extensions\ms-python.debugpy-2024.10.0-win32-x64\bundled\libs\debugpy\adapter/ ... /... \debugpy\launcher' "51870" "--" "C:\Users\huang\Documents\GitHub\training\scripts\test.py"
Token found: githu
Repository successfully accessed: NFDI4BIOIMAGE/training
PS C:\Users\huang\Documents\GitHub\training\scripts> 

I'm trying to see if this can be avoided, but nothing I've seen so far seems to work! Or maybe not with github token, I'm not sure

**Comment by haesleinhuepf** [comment]
> nothing I've seen so far seems to work!

Yes, it is possible that this will not work. Do you have an overview about which meta data we can access and which not?

**Comment by SeverusYixin** [comment]
> > nothing I've seen so far seems to work!
> 
> Yes, it is possible that this will not work. Do you have an overview about which meta data we can access and which not?

At the moment, it looks like the request will be denied when you try to access it with github token.

**Comment by haesleinhuepf** [comment]
Can you please try "repo.get_contributors()" instead of "repo.get_collaborators()"?

**Comment by SeverusYixin** [comment]
> Can you please try "repo.get_contributors()" instead of "repo.get_collaborators()"?

Let me try

**Comment by haesleinhuepf** [comment]
Did it work?

**Comment by SeverusYixin** [comment]
> Did it work?

no

**Comment by SeverusYixin** [comment]
Please do not merge. There are still some details to discuss :)

1. Currently it looks like #238, so again, what should the name be, "ymia_python_event_series_material" or "YMIA - Python-Based Event Series Training Material"?

2. and for tags, do we need this "tags: No tags available"? or this "tags:   "

**Comment by haesleinhuepf** [comment]
> 1. "ymia_python_event_series_material" or "YMIA - Python-Based Event Series Training Material"?

I would use whatever the github returns as `repo.name`

> 2\. and for tags, do we need this "tags: No tags available"? or this "tags:   "

If there are no tags, we should not have  a "tag:" in our yaml.

**Comment by haesleinhuepf** [comment]
> no

Consider posting an error message, in case you wish that I assist you.

**Comment by SeverusYixin** [comment]
> Consider posting an error message, in case you wish that I assist you.

Already solved, thank you if I need that I will post a error message. Thank you again :)

**Comment by haesleinhuepf** [comment]
Are there any news here @SeverusYixin ? Shall I test it agian?

**Comment by SeverusYixin** [comment]
> Are there any news here @SeverusYixin ? Shall I test it agian?

Seems like you didn't see issues #185.

I asked you how should we define the type and tag for a github repositorie. Except for type and day, all other functions are implemented.

**Comment by SeverusYixin** [comment]
write something for the tags like "Dear users, please add a tags"

**Comment by SeverusYixin** [comment]
And never have empty strings

**Comment by SeverusYixin** [comment]
Hi @haesleinhuepf, please review again, if everything is ok then we can close and merge this.

Thank you :)

File: scripts/auto-add-github-resources.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+    # Repository description
+    entry['description'] = repo.description if repo.description else ""
+
+    # License information (stored as "license")
+    try:
+        license_info = repo.get_license()
+        entry['license'] = license_info.license.name if license_info else ""
+    except:
+        entry['license'] = ""
```
Comment (haesleinhuepf) [inline scripts/auto-add-github-resources.py:95]:
```suggestion
        pass
```

File: scripts/auto-add-github-resources.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        else:
+            raise e
+
+    # Repository description
+    entry['description'] = repo.description if repo.description else ""
+
+    # License information (stored as "license")
+    try:
+        license_info = repo.get_license()
+        entry['license'] = license_info.license.name if license_info else ""
```
Comment (haesleinhuepf) [inline scripts/auto-add-github-resources.py:93]:
```suggestion
        if license_info is not None:
            entry['license'] = license_info.license.name
```

File: scripts/auto-add-github-resources.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    # License information (stored as "license")
+    try:
+        license_info = repo.get_license()
+        if license_info is not None:
+            entry['license'] = license_info.license.name
+    except:
+        pass
+
+    # Repository name
+    entry['name'] = repo.name if repo.name else ""
```
Comment (haesleinhuepf) [inline scripts/auto-add-github-resources.py:99]:
```suggestion
    entry['name'] = repo.name
```

File: scripts/auto-add-github-resources.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+    Returns:
+    - publication_date (str): The ISO-formatted publication date.
+    """
+    releases = repo.get_releases()
+    if releases.totalCount > 0:
+        first_release = releases[0]
+        return first_release.created_at.isoformat()
+    else:
+        return repo.created_at.isoformat() if repo.created_at else ""
```
Comment (haesleinhuepf) [inline scripts/auto-add-github-resources.py:137]:
```suggestion
    elif repo.created_at is not None:
        return repo.created_at.isoformat()
```

File: scripts/auto-add-github-resources.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    # Publication date (first release date or creation date)
+    entry['publication_date'] = get_publication_date(repo)
+
+    # Tags: always add the message "Dear users, please add tags."
+    entry['tags'] = """This is an automatically created pull-request. Before merging this, please make sure
+* [ ] has proper `tags` set.
+* [ ] has reasonable content `type` depending on the content of the github repository.
+
+ [Read more](https://nfdi4bioimage.github.io/training/contributing/format.html) 
+    """
```
Comment (haesleinhuepf) [inline scripts/auto-add-github-resources.py:110]:
```suggestion
    entry['tags'] = "TODO"
```

File: scripts/auto-add-github-resources.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    file_content = get_file_in_repository(repository, branch, yml_filename).decoded_content.decode()
+
+    print("YAML file content length:", len(file_content))
+
+    # Add the new entry
+    file_content += github_yml
+
+    # Save the updated content back to GitHub and create a pull request
+    write_file(repository, branch, yml_filename, file_content, "Add " + github_repo_url)
+    res = send_pull_request(repository, branch, "Add " + github_repo_url, f"closes #{issue}")
```
Comment (haesleinhuepf) [inline scripts/auto-add-github-resources.py:48]:
```suggestion
    res = send_pull_request(repository, branch, "Add " + github_repo_url, f"""This is an automatically created pull-request. Before merging this, please make sure
* [ ] has proper `tags` set.
* [ ] has reasonable content `type` depending on the content of the github repository.
 [Read more](https://nfdi4bioimage.github.io/training/contributing/format.html) 
   
closes #{issue}
""")
```

**Pull request merged**